### PR TITLE
MetroDR: clear PV ClaimRef on secondary PVC deletion and put back the ClaimRef on restore

### PR DESCRIPTION
**Pull request by BenamarMk**
In MetroDR we don't change the reclaim policy for the PV from `Retain` back to `Delete` when we failover/relocate an app to another cluster. In other words, if we failover from C2 to C1, then the PV on C2 will be retained. Therefore, the PV will not be deleted but its Status.Phase will be move to `Released`.

In that case, if we failover or relocate back to that cluster, then it is guaranteed that a new PV will be created instead of binding to the released one.

Kubernetes will only bind a PVC to a PV if the PV is in the `Available` status phase.

The fix in this PR is to clear the ClaimRef from the PV when the PVC is deleted. On Restore, if we find that the PV exists, then we update its ClaimRef with the ClaimRef of the PV that we download from the s3 store.

### CODE REVIEW

**Comment by BenamarMk** [comment]
@raghavendra-talur Please add the unit test that you wanted to add to validate this.

**Review by raghavendra-talur — APPROVED** [review]

File: controllers/vrg_volrep.go [hunk 1]
```diff
@@ -318,9 +318,14 @@ func (v *VRGInstance) preparePVCForVRDeletion(pvc *corev1.PersistentVolumeClaim,
...
 	// both the primary and the secondary VRG map to the same volume. The only
 	// state where a delete retention policy is required for the sync mode is
-	// when the VRG is primary.
-	if !(v.instance.Spec.ReplicationState == ramendrv1alpha1.Secondary &&
-		v.instance.Spec.Sync.Mode == ramendrv1alpha1.SyncModeEnabled) {
+	// when the VRG is primary. Furthermore, we need to clear the PV claimRef
+	// in order for the PV to go to the Available status phase.
+	if v.instance.Spec.ReplicationState == ramendrv1alpha1.Secondary &&
+		v.instance.Spec.Sync.Mode == ramendrv1alpha1.SyncModeEnabled {
+		if err := v.clearPVClaimRef(pvc); err != nil {
```
Comment (ShyamsundarR) [inline controllers/vrg_volrep.go:325]: Technically the PV is now Available for another PVC request of the same size and StorageClass, the k8s provisioner hence may bind this PV to the new PVC request in another namespace and leak storage across namespaces.

Afaict, all Available PVs are looked at and if any satisfies the request it is bound to the new PVC request.

I do not have an alternative to suggest though, but we may want to track this caveat.

Comment (BenamarMk) [reply controllers/vrg_volrep.go:325]: That's true. We can put a dummy PVC name there as in Metro and in this PR, we don't check whether ClaimRef is nil or not.  We blindly assign to it the ClaimRef from the PV that we downloaded from the s3 store.  So if we stick in the ClaimRef of the existing PV a random name/namespace, like GUID, then that would solve the issue.

Comment (raghavendra-talur) [reply controllers/vrg_volrep.go:325]: What if we just remove the UUID and the resource version of the claimref. That way we ensure that the name of the pvc matches. I tried this on a minikube instance and it worked.

Comment (BenamarMk) [reply controllers/vrg_volrep.go:325]: @raghavendra-talur, I think that should work, but I was thinking something like this
```
pv.Spec.ClaimRef = &corev1.ObjectReference {
		Name: pv.Spec.ClaimRef.Name + pv.Spec.ClaimRef.ResourceVersion,
		Namespace: pv.Spec.ClaimRef.Namespace + pv.Spec.ClaimRef.ResourceVersion,
	}
```

Comment (BenamarMk) [reply controllers/vrg_volrep.go:325]: But if we only remove the UID and the Resource version, then it should work without the changes that I made for restore, does that sound right?

Comment (raghavendra-talur) [reply controllers/vrg_volrep.go:325]: Yes, no changes required during the restore.

Comment (BenamarMk) [reply controllers/vrg_volrep.go:325]: Ok, I am going to leave this PR open, we can close it later, but I will create a different PR with only the changes we talked about.

**Pull request closed without merging**